libp2p/module.go

```diff
@@ -0,0 +1,367 @@
+		mod.listenAddrs = libp2p.NoListenAddrs
+	default:
+		return nil, fmt.Errorf(
+			// DISCUSS: rename to "transport protocol" instead.
+			"unsupported connection type: %s: %w",
+			mod.cfg.ConnectionType,
+			err,
+		)
+	}
+
+	return mod, nil
+}
+
+func (mod *libp2pModule) Start() error {
+	mod.logger = logger.Global.CreateLoggerForModule(mod.GetModuleName())
+
+	// IMPROVE: receive context in interface methods.
+	ctx := context.Background()
+
+	// TECHDEBT: metrics integration.
+	var err error
+	opts := []libp2p.Option{
+		mod.identity,
+		// INCOMPLETE(#544): add transport security!
+	}
+
+	// Disable unused libp2p relay and ping services in client debug mode.
+	// (see: https://pkg.go.dev/github.com/libp2p/go-libp2p#DisableRelay
+	// and https://pkg.go.dev/github.com/libp2p/go-libp2p#Ping)
+	if mod.GetBus().GetRuntimeMgr().GetConfig().ClientDebugMode {
+		opts = append(opts,
+			libp2p.DisableRelay(),
+			libp2p.Ping(false),
+			libp2p.NoListenAddrs,
+		)
+	} else {
+		opts = append(opts, mod.listenAddrs)
+	}
+
+	// Represents a libp2p network node, `libp2p.New` configures
+	// and starts listening according to options.
+	// (see: https://pkg.go.dev/github.com/libp2p/go-libp2p#section-readme)
+	mod.host, err = libp2p.New(opts...)
+	if err != nil {
+		return fmt.Errorf("unable to create libp2p host: %w", err)
+	}
+
+	listenAddrLogEvent := mod.logger.Info()
+	for i, addr := range host.InfoFromHost(mod.host).Addrs {
+		listenAddrLogEvent.Str(fmt.Sprintf("listen_addr_%d", i), addr.String())
+	}
+	listenAddrLogEvent.Msg("Listening for incoming connections...")
+
+	// TECHDEBT: use RandomSub or GossipSub once we're on more stable ground.
+	// IMPROVE: consider supporting multiple router types via config.
+	mod.pubsub, err = pubsub.NewFloodSub(ctx, mod.host)
+	if err != nil {
+		return fmt.Errorf("unable to create pubsub: %w", err)
+	}
+
+	// Topic is used to `#Publish` messages.
+	mod.topic, err = mod.pubsub.Join(protocol.DefaultTopicStr)
+	if err != nil {
+		return fmt.Errorf("unable to join pubsub topic: %w", err)
+	}
+
+	// Subscription is notified when a new message is received on the topic.
+	mod.subscription, err = mod.topic.Subscribe()
+	if err != nil {
+		return fmt.Errorf("subscribing to pubsub topic: %w", err)
+	}
+
+	mod.network, err = network.NewLibp2pNetwork(mod.GetBus(), mod.logger, mod.host, mod.topic)
+	if err != nil {
+		return fmt.Errorf("creating network: %w", err)
+	}
+
+	// Don't handle streams or read from the subscription in client debug mode.
+	if !mod.GetBus().GetRuntimeMgr().GetConfig().ClientDebugMode {
+		mod.host.SetStreamHandler(protocol.PoktProtocolID, mod.handleStream)
+		go mod.readFromSubscription(ctx)
+	}
+	return nil
+}
+
+func (mod *libp2pModule) Stop() error {
+	return mod.host.Close()
+}
+
+func (mod *libp2pModule) Broadcast(msg *anypb.Any) error {
+	c := &messaging.PocketEnvelope{
+		Content: msg,
+	}
+	data, err := proto.MarshalOptions{Deterministic: true}.Marshal(c)
```

Comment:
We have some techdebt here (my fault) that you ported from P2P, would you mind adding a 
`\\TECHDEBT: use shared/codec for marshalling` here, below in the `Send` and if you don't mind also in the same methods in `p2p` module?